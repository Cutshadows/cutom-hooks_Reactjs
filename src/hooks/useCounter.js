import {useState} from 'react';


export const useCounter=(initialState=10)=>{
    const [counter, setState] = useState(initialState)


    const increment=(factor=1)=>{
        setState(counter+factor)
    }

    const decrementar=(factor=1)=>{
        setState(counter-factor)
    }

    const reset=()=>{
        setState(initialState)
    }

    return {
        counter,
        increment,
        decrementar,
        reset
    };
}