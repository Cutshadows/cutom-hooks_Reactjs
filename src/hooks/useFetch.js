import { useEffect, useRef, useState } from "react";

const useFetch = (url) => {
    
    const isMounted = useRef(true);
    
    const [state, setState] = useState({
        data:null,
        loading:true,
        error:null
    })
    useEffect(() => {
        return()=>{
            isMounted.current=false;
        }
    }, [])

    useEffect(() => {
        setState({
            data:null,
            loading:true,
            error:null
        })
        fetch(url)
            .then(resp=>resp.json())
            .then(data=>{
                    // setTimeout(()=>{
                        if(isMounted.current){
                                setState({
                                    loading:false,
                                    error:null,
                                    data
                                })
                        }
                    // }, 3000)
            }).catch(()=>{
                setState({
                    data:null,
                    loading:false,
                    error:'No se pudo cargar la informaci√≥n'
                })  
            })
    }, [url])

    return state;

}

//https://www.breakingbadapi.com/api/
 
export default useFetch;