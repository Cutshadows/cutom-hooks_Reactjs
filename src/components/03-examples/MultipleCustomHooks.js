import React from 'react';
import useFetch from '../../hooks/useFetch';
import {useCounter} from '../../hooks/useCounter';
import './multiplecustomhooks.scss';

const MultipleCustomHooks = () => {

    const {increment, state}=useCounter(1);
    const {loading, data}=useFetch(`https://www.breakingbadapi.com/api/quotes/${state}`);

    const {author, quote}=!!data&&data[0]; //condicion logica que se evalua con doble negacion quiero transformar ese null en valor booleano

    return ( 
        <div> 
            <h1>MultipleCustomHooks</h1>
            
            {loading?
            (
                <div className="alert alert-info text-center">
                loading...
                </div>
            ):
            (
                <blockquote className="blockquote text-right">
                    <p className="mb-0">{quote}</p>
                    <footer className="blockquote-footer">{author}</footer>
                </blockquote>
            )
            }
            
            
            <button className="btn btn-primary"
            onClick={()=>increment()}>
                Siguiente quote
            </button>
            

         </div>
     );
}
 
export default MultipleCustomHooks;