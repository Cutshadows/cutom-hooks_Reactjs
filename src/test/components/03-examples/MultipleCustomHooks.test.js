import {act, renderHook} from '@testing-library/react-hooks';
import { shallow } from 'enzyme';
import MultipleCustomHooks from '../../../components/03-examples/MultipleCustomHooks';
import { useCounter } from '../../../hooks/useCounter';
import useFetch from '../../../hooks/useFetch';

jest.mock('../../../hooks/useFetch');
jest.mock('../../../hooks/useCounter');


describe('Pruebas de <MultipleCustomHooks></MultipleCustomHooks>', () => {
    beforeEach(()=>{
        useCounter.mockReturnValue({
            counter:10,
            increment:()=>{}
        })
    })

    test('debe de mostrar correctamente', () => {
        useFetch.mockReturnValue({      
            data:null,
            loading:true,
            error:null
        })
        const wrapper=shallow(<MultipleCustomHooks/>);
        expect(wrapper).toMatchSnapshot();
        
    })
    test('debe de mostrar la informacion', () => {
        useFetch.mockReturnValue({
            data:[{
                author:'Fernando',
                quote:'Hola Douglas'
            }],
            loading:false,
            error:null
        });

        const wrapper=shallow(<MultipleCustomHooks/>);
        expect(wrapper.find('.alert').exists()).toBe(false);
        expect(wrapper.find('.mb-0').text().trim()).toBe('Hola Douglas');
        expect(wrapper.find('.blockquote-footer').text().trim()).toBe('Fernando');
        expect(wrapper.find('footer').text().trim()).toBe('Fernando');
        // console.log(wrapper.html())

    })
    
})
