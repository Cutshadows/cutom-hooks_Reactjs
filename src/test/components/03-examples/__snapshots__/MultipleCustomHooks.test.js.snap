// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas de <MultipleCustomHooks></MultipleCustomHooks> debe de mostrar correctamente 1`] = `
<div>
  <h1>
    MultipleCustomHooks
  </h1>
  <div
    className="alert alert-info text-center"
  >
    loading...
  </div>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Siguiente quote
  </button>
</div>
`;
